kernel-switcher/test_in_lxd_vm.py

```diff
@@ -0,0 +1,155 @@
+
+        self._run_host_command(
+            f"lxc launch ubuntu:22.04 {self.vm_name} --vm -c security.secureboot=false"
+        )
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        print(f"Deleting '{self.vm_name}' VM...")
+        self._run_host_command(f"lxc delete --force {self.vm_name}")
+
+    def _run_host_command(self, cmd, supress=False):
+        """Executes a command on the host and returns the output."""
+        try:
+            cmd_list = shlex.split(cmd)
+            result = subprocess.run(
+                cmd_list,
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                text=True,
+            )
+            return result.stdout
+        except subprocess.CalledProcessError as exc:
+            if not supress:
+                print(f"Error while running command: {cmd}")
+                print(f"Error message: {exc.output}")
+            raise
+
+    def run_vm_command(self, cmd, supress=False):
+        """Executes a command inside the VM and returns the output."""
+        print(f"Running command inside VM: {cmd}")
+        full_cmd = f"lxc exec {self.vm_name} -- {cmd}"
+        return self._run_host_command(full_cmd, supress)
+
+    def wait_for_vm_ready(self, timeout=60):
+        """Waits until the VM is responsive."""
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            try:
+                self.run_vm_command("true", supress=True)
+                return
+            except subprocess.CalledProcessError:
+                time.sleep(1)
+        raise TimeoutError(
+            f"VM '{self.vm_name}' did not become ready within {timeout} seconds."
+        )
+
+
+def extract_kernel_pkgs_from_apt_log(apt_log):
+    """
+    Returns the name of the kernel package from the apt log.
+
+    >>> text = '''
+    ... linux-image-5.4.0-80-generic/focal-updates,rubbish [more rubbish]]]
+    ... linux-image-5.4.0-81-generic/focal-updates, no rubbish
+    ... at all!
+    ... well almost, okay, one more linux-image like incoming
+    ... linux-image-5.15.0-1041-kvm/jammy-updates 5.15.0-1041.46 amd64'''
+    >>> extract_kernel_pkgs_from_apt_log(text) == [
+    ...     'linux-image-5.4.0-80-generic',
+    ...     'linux-image-5.4.0-81-generic',
+    ...     'linux-image-5.15.0-1041-kvm']
+    True
+    """
+    # list comprehension here would be unreadable due to line length
+    # let's do this imperatively
+    # not using a regex because even here it may be easier to understand
+    # with just python's string methods
+    extraceted_pkgs = []
+
+    for line in apt_log.splitlines():
+        if line.startswith("linux-image"):
+            extraceted_pkgs.append(line.split("/")[0])
+    return extraceted_pkgs
+
+
+def main():
+    interrupt_handler = InterruptHandler()
+    with VMContext("vm-test") as vm:
+        # Wait for VM to be ready
+        time.sleep(10)
```

Comment:
is this equivalent to `vm.wait_for_vm_ready(timeout=70)`

Reply:
Not fully equivalent... The time.sleep() is not an active wait so nothing is run and being printed. But with a testing tool like this it's not necessary to have super clean output.
Deleted.